client/cli/src/lib.rs

```diff
@@ -177,461 +179,393 @@ fn is_node_name_valid(_name: &str) -> Result<(), &str> {
+
+	let mut full_version = sc_service::config::full_version_from_strs(
 		version.version,
 		version.commit
 	);
+	full_version.push_str("\n");
 
-	sp_panic_handler::set(version.support_url, &full_version);
-	let matches = CoreParams::<CC, RP>::clap()
+	let app = app
 		.name(version.executable_name)
 		.author(version.author)
 		.about(version.description)
-		.version(&(full_version + "\n")[..])
-		.setting(AppSettings::GlobalVersion)
-		.setting(AppSettings::ArgsNegateSubcommands)
-		.setting(AppSettings::SubcommandsNegateReqs)
-		.get_matches_from(args);
-	let cli_args = CoreParams::<CC, RP>::from_clap(&matches);
-	fdlimit::raise_fd_limit();
+		.version(full_version.as_str());
 
-	let args = match cli_args {
-		params::CoreParams::Run(params) => ParseAndPrepare::Run(
-			ParseAndPrepareRun { params, impl_name, version }
-		),
-		params::CoreParams::BuildSpec(params) => ParseAndPrepare::BuildSpec(
-			ParseAndPrepareBuildSpec { params, version }
-		),
-		params::CoreParams::ExportBlocks(params) => ParseAndPrepare::ExportBlocks(
-			ParseAndPrepareExport { params, version }
-		),
-		params::CoreParams::ImportBlocks(params) => ParseAndPrepare::ImportBlocks(
-			ParseAndPrepareImport { params, version }
-		),
-		params::CoreParams::CheckBlock(params) => ParseAndPrepare::CheckBlock(
-			CheckBlock { params, version }
-		),
-		params::CoreParams::PurgeChain(params) => ParseAndPrepare::PurgeChain(
-			ParseAndPreparePurge { params, version }
-		),
-		params::CoreParams::Revert(params) => ParseAndPrepare::RevertChain(
-			ParseAndPrepareRevert { params, version }
-		),
-		params::CoreParams::Custom(params) => ParseAndPrepare::CustomCommand(params),
-	};
-	init_logger(args.shared_params().and_then(|p| p.log.as_ref()).map(|v| v.as_ref()).unwrap_or(""));
-	args
+	T::from_clap(&app.get_matches_from(iter))
 }
 
-/// Returns a string displaying the node role, special casing the sentry mode
-/// (returning `SENTRY`), since the node technically has an `AUTHORITY` role but
-/// doesn't participate.
-pub fn display_role<A, B, C>(config: &Configuration<A, B, C>) -> String {
-	if config.sentry_mode {
-		"SENTRY".to_string()
-	} else {
-		format!("{:?}", config.roles)
-	}
-}
-
-/// Output of calling `parse_and_prepare`.
-#[must_use]
-pub enum ParseAndPrepare<'a, CC, RP> {
-	/// Command ready to run the main client.
-	Run(ParseAndPrepareRun<'a, RP>),
-	/// Command ready to build chain specs.
-	BuildSpec(ParseAndPrepareBuildSpec<'a>),
-	/// Command ready to export the chain.
-	ExportBlocks(ParseAndPrepareExport<'a>),
-	/// Command ready to import the chain.
-	ImportBlocks(ParseAndPrepareImport<'a>),
-	/// Command to check a block.
-	CheckBlock(CheckBlock<'a>),
-	/// Command ready to purge the chain.
-	PurgeChain(ParseAndPreparePurge<'a>),
-	/// Command ready to revert the chain.
-	RevertChain(ParseAndPrepareRevert<'a>),
-	/// An additional custom command passed to `parse_and_prepare`.
-	CustomCommand(CC),
-}
+pub fn run<F, G, E, F2, F3, F4, T1, T2, T3, T5>(
```

Comment:
You make these traits a little more descriptive, like the way they are in `Service`.

Reply:
Yes :grin: this was temporary. Check again, I removed the numbers